source/org/zfin/expression/repository/HibernateExpressionRepository.java

```diff
@@ -63,1372 +59,1296 @@
+		});
+	}
+
+	/**
+	 * Delete a expressionFigureStage record including all expression results on it
+	 *
+	 * @param figureAnnotation experiment figure stage.
+	 */
+	public void deleteFigureAnnotation(ExpressionFigureStage figureAnnotation) {
+		if (figureAnnotation == null) {
+			throw new NullPointerException("No Figure Annotation provided");
+		}
+		HibernateUtil.currentSession().delete(figureAnnotation);
+	}
+
+	/**
+	 * Retrieve an efs by experiment, figure, start and end stage id.
+	 *
+	 * @param experimentZdbID experiment
+	 * @param figureID        figure
+	 * @param startStageID    start
+	 * @param endStageID      end
+	 * @return efs object
+	 */
+	public ExpressionFigureStage getExperimentFigureStage(String experimentZdbID, String figureID, String startStageID, String endStageID) {
+		if (StringUtils.isEmpty(experimentZdbID)) {
+			return null;
+		}
+		if (StringUtils.isEmpty(figureID)) {
+			return null;
+		}
+		if (StringUtils.isEmpty(startStageID)) {
+			return null;
+		}
+		if (StringUtils.isEmpty(endStageID)) {
+			return null;
+		}
+		validateFigureAnnotationKey(experimentZdbID, figureID, startStageID, endStageID);
+
+		Session session = HibernateUtil.currentSession();
+
+		String hql = "select efs from ExpressionFigureStage as efs ";
+		hql += "     where efs.expressionExperiment.zdbID = :experimentID ";
+		hql += " AND efs.startStage.zdbID = :startID ";
+		hql += " AND efs.endStage.zdbID = :endID ";
+		hql += " AND efs.figure.zdbID = :figureID ";
+		Query<ExpressionFigureStage> query = session.createQuery(hql, ExpressionFigureStage.class);
+		query.setParameter("experimentID", experimentZdbID);
+		query.setParameter("startID", startStageID);
+		query.setParameter("endID", endStageID);
+		query.setParameter("figureID", figureID);
+
+		return query.uniqueResult();
+	}
+
+	/**
+	 * Retrieve all expression structures for a given publication, which is the same as the
+	 * structure pile.
+	 *
+	 * @param publicationID publication ID
+	 * @return list of expression structures.
+	 */
+	public List<ExpressionStructure> retrieveExpressionStructures(String publicationID) {
+		Session session = HibernateUtil.currentSession();
+		Criteria crit = session.createCriteria(ExpressionStructure.class);
+		crit.add(Restrictions.eq("publication.zdbID", publicationID));
+		Criteria superterm = crit.createCriteria("superterm");
+		superterm.addOrder(Order.asc("termName"));
+		crit.setFetchMode("superterm", FetchMode.JOIN);
+		crit.setFetchMode("superterm.start", FetchMode.JOIN);
+		crit.setFetchMode("superterm.end", FetchMode.JOIN);
+		crit.setFetchMode("person", FetchMode.JOIN);
+		return (List<ExpressionStructure>) crit.list();
+	}
+
+	/**
+	 * Retrieve a single expression structure by ID.
+	 *
+	 * @param zdbID structure ID
+	 * @return expression structure
+	 */
+	public ExpressionStructure getExpressionStructure(String zdbID) {
+		Session session = HibernateUtil.currentSession();
+		return session.get(ExpressionStructure.class, zdbID);

```

Comment:
Is there any potential for edge cases where ".get" returns null and ".uniqueResult" throws exception?

Reply:
Not fully sure but:
returning a null should be ok. the calling method should handle the null.
and the exception being thrown when uniqueResult will retrieve multiple records is fine as is. If the result set should always be unique this exception will fastest show if it is not true in some cases. I definitely would not call uniqueResult() knowing that there are cases where it is not true. In that case, you should call list() and then check according to best business logic how to handle the violation. if you need to check if something has a unique resultset you better call list() and then return the boolean and then proceed with both logical pathways from there. You typically should not use exceptions to go one or the other way in your code.